### ISSUE TITLE
Bootstrapping with Redis/LPUSH fails on all but tiny tables

### ISSUE THREAD

**Issue by blockmar** [zendesk/maxwell#1161]
Using the latest maxwell version (1.19.1) and trying to bootstrap tables using the redis provider with lpush.

If I try a small table with one about 150 rows it works well but if a try a bigger table with about 70K rows maxwell fails to write to Redis due to a broken pipe.

The exception is repeated for every row in the table. I think it fails from the first "bootstrap-insert". I can verify that "bootstrap-start".

Rested with both Redis 3 and 4.

Normal replication and bootstrapping of small tables works fine.

```
Using kafka version: 1.0.0
20:49:50,536 WARN  MaxwellMetrics - Metrics will not be exposed: metricsReportingType not configured.
20:49:51,114 INFO  Maxwell - Maxwell v1.19.1 is booting (MaxwellRedisProducer), starting at Position[BinlogPosition[mysql-bin-changelog.017173:104839458], lastHeartbeat=1542228414251]
20:49:51,298 INFO  MysqlSavedSchema - Restoring schema id 1 (last modified at Position[BinlogPosition[mysql-bin-changelog.016831:108633155], lastHeartbeat=0])
20:49:51,536 INFO  BinlogConnectorReplicator - Setting initial binlog pos to: mysql-bin-changelog.017173:104839458
20:49:51,685 INFO  BinaryLogClient - Connected to xxxxxxxxxxxxx:3306 at mysql-bin-changelog.01XXX3/104XXXX58 (sid:6379, cid:181381)
20:49:51,690 INFO  BinlogConnectorLifecycleListener - Binlog connected.
20:50:12,967 ERROR MaxwellRedisProducer - Exception during put
redis.clients.jedis.exceptions.JedisConnectionException: java.net.SocketException: Broken pipe (Write failed)
	at redis.clients.jedis.Connection.flush(Connection.java:334) ~[jedis-2.9.0.jar:?]
	at redis.clients.jedis.Connection.getIntegerReply(Connection.java:263) ~[jedis-2.9.0.jar:?]
	at redis.clients.jedis.Jedis.lpush(Jedis.java:882) ~[jedis-2.9.0.jar:?]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.push(MaxwellRedisProducer.java:45) [maxwell-1.19.1.jar:1.19.1]
	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.processRow(BinlogConnectorReplicator.java:207) [maxwell-1.19.1.jar:1.19.1]
	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.work(BinlogConnectorReplicator.java:162) [maxwell-1.19.1.jar:1.19.1]
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:27) [maxwell-1.19.1.jar:1.19.1]
	at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:224) [maxwell-1.19.1.jar:1.19.1]
	at com.zendesk.maxwell.Maxwell.start(Maxwell.java:156) [maxwell-1.19.1.jar:1.19.1]
	at com.zendesk.maxwell.Maxwell.main(Maxwell.java:245) [maxwell-1.19.1.jar:1.19.1]
Caused by: java.net.SocketException: Broken pipe (Write failed)
	at java.net.SocketOutputStream.socketWrite0(Native Method) ~[?:?]
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111) ~[?:?]
	at java.net.SocketOutputStream.write(SocketOutputStream.java:155) ~[?:?]
	at redis.clients.util.RedisOutputStream.flushBuffer(RedisOutputStream.java:52) ~[jedis-2.9.0.jar:?]
	at redis.clients.util.RedisOutputStream.flush(RedisOutputStream.java:216) ~[jedis-2.9.0.jar:?]
	at redis.clients.jedis.Connection.flush(Connection.java:331) ~[jedis-2.9.0.jar:?]
	... 9 more
```

**Comment by lucastex** [comment]
Hi @osheroff .  I keep seeing this problem when bootstrapping large tables. There is not actually a pattern on this, sometimes it bootstrap a 20k table, sometimes it doesn't on a 1k table.

I saw the reconnect logic implemented, but it is failing on the reconnection since it just can't connect with redis.

Any ideas on this? I'm thinking the problem is probably on the jedis internal commands. I'll try to implement the RedisProducer with Redisson, a high performance redis lib and test it locally to see if I see any changes.

**Comment by osheroff** [comment]
Does it time out trying to connect?  Is red is cpu-bound or crashing?  Or maybe jedis has got its internal state fucked up and so can’t reconnect?

More error logs would be helpful

**Comment by lucastex** [comment]
1 - It connects and starts to send data, and then, suddenly it just disconnects, and fail to reconnect. 

2 - All the times this happened, Redis CPU wasn't higher than 5%.

At first looking into it looks like a connectivity issue between maxwell and redis, but doesn't make sense, because if we wait *a lot*, the connectivity won't be restabilshed, and if we just restart maxwell everything would work normally. 

Oh, and we have other maxwells sending to the same redis (and sending from the same server), so, one is sending ok, while the other can't connect.

I'd risk on something on jedis internals. Do you mind if I make some tests using other redis lib?
I can try to work on some tests for it.

Thanks @osheroff

**Comment by osheroff** [comment]
go for it.  i still think you should post your error logs.

**Comment by lucastex** [comment]
Oh, sorry, didn't send in the previous message. This is what I just reproduced. 
Did it just running a full bootstrap on my table, and it started to crash in the 60kth row.

```23:27:25,959 WARN  MaxwellRedisProducer - lost connection to server, trying to reconnect...
redis.clients.jedis.exceptions.JedisConnectionException: Attempting to read from a broken connection
	at redis.clients.jedis.Connection.readProtocolWithCheckingBroken(Connection.java:314) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.Connection.getIntegerReply(Connection.java:260) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.Jedis.lpush(Jedis.java:977) ~[jedis-3.1.0.jar:?]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.sendToRedis(MaxwellRedisProducer.java:61) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.push(MaxwellRedisProducer.java:124) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.SynchronousBootstrapper.performBootstrap(SynchronousBootstrapper.java:81) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.SynchronousBootstrapper.startBootstrap(SynchronousBootstrapper.java:42) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.BootstrapController.doWork(BootstrapController.java:74) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.BootstrapController.work(BootstrapController.java:58) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:34) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.MaxwellContext.lambda$getBootstrapController$0(MaxwellContext.java:400) ~[maxwell-1.25.0.jar:1.25.0]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_242]
23:27:26,035 WARN  MaxwellRedisProducer - lost connection to server, trying to reconnect...
redis.clients.jedis.exceptions.JedisConnectionException: Attempting to read from a broken connection
	at redis.clients.jedis.Connection.readProtocolWithCheckingBroken(Connection.java:314) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.Connection.getIntegerReply(Connection.java:260) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.Jedis.lpush(Jedis.java:977) ~[jedis-3.1.0.jar:?]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.sendToRedis(MaxwellRedisProducer.java:61) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.push(MaxwellRedisProducer.java:124) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.SynchronousBootstrapper.performBootstrap(SynchronousBootstrapper.java:81) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.SynchronousBootstrapper.startBootstrap(SynchronousBootstrapper.java:42) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.BootstrapController.doWork(BootstrapController.java:74) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.BootstrapController.work(BootstrapController.java:58) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:34) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.MaxwellContext.lambda$getBootstrapController$0(MaxwellContext.java:400) ~[maxwell-1.25.0.jar:1.25.0]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_242]
23:27:26,036 WARN  MaxwellRedisProducer - lost connection to server, trying to reconnect...
redis.clients.jedis.exceptions.JedisConnectionException: Attempting to read from a broken connection
	at redis.clients.jedis.Connection.readProtocolWithCheckingBroken(Connection.java:314) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.Connection.getIntegerReply(Connection.java:260) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.Jedis.lpush(Jedis.java:977) ~[jedis-3.1.0.jar:?]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.sendToRedis(MaxwellRedisProducer.java:61) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.push(MaxwellRedisProducer.java:124) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.SynchronousBootstrapper.performBootstrap(SynchronousBootstrapper.java:81) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.SynchronousBootstrapper.startBootstrap(SynchronousBootstrapper.java:42) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.BootstrapController.doWork(BootstrapController.java:74) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.bootstrap.BootstrapController.work(BootstrapController.java:58) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:34) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.MaxwellContext.lambda$getBootstrapController$0(MaxwellContext.java:400) ~[maxwell-1.25.0.jar:1.25.0]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_242]```

**Comment by lucastex** [comment]
As I said, everything makes it looks like a network issue, but as I said, we have other maxwell instances in the same machine, sending to the same redis.

**Comment by osheroff** [comment]
yeah that's definitely internal jedis state corrupted.  do you have the *first* error message that happened on the node?

**Comment by lucastex** [comment]
The log was set to WARN in this execution. I'll change it and get more information tomorrow and post here.

**Comment by osheroff** [comment]
hmmm.  I'm thinking there should be an initial exception, different from "Attempting to read from a broken connection".  Am I wrong?

**Comment by osheroff** [comment]
the Maxwell warning message should be the same but I'm guess there's a different redis exception underlying?

**Comment by lucastex** [comment]
Root cause is `socket is not connected` @osheroff 

```23:27:27,735 ERROR TaskManager - cause: 
redis.clients.jedis.exceptions.JedisConnectionException: Failed connecting to host xxxxx:6379
	at redis.clients.jedis.Connection.connect(Connection.java:204) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.BinaryClient.connect(BinaryClient.java:100) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.BinaryJedis.connect(BinaryJedis.java:1866) ~[jedis-3.1.0.jar:?]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.push(MaxwellRedisProducer.java:130) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.processRow(BinlogConnectorReplicator.java:280) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.work(BinlogConnectorReplicator.java:193) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:34) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:223) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.Maxwell.start(Maxwell.java:156) ~[maxwell-1.25.0.jar:1.25.0]
	at com.zendesk.maxwell.Maxwell.main(Maxwell.java:244) ~[maxwell-1.25.0.jar:1.25.0]
Caused by: java.net.SocketException: Socket is not connected
	at java.net.Socket.getOutputStream(Socket.java:963) ~[?:1.8.0_242]
	at redis.clients.jedis.Connection.connect(Connection.java:200) ~[jedis-3.1.0.jar:?]
	... 9 more
redis.clients.jedis.exceptions.JedisConnectionException: Failed connecting to host xxxxx:6379
	at redis.clients.jedis.Connection.connect(Connection.java:204)
	at redis.clients.jedis.BinaryClient.connect(BinaryClient.java:100)
	at redis.clients.jedis.BinaryJedis.connect(BinaryJedis.java:1866)
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.push(MaxwellRedisProducer.java:130)
	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.processRow(BinlogConnectorReplicator.java:280)
	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.work(BinlogConnectorReplicator.java:193)
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:34)
	at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:223)
	at com.zendesk.maxwell.Maxwell.start(Maxwell.java:156)
	at com.zendesk.maxwell.Maxwell.main(Maxwell.java:244)
Caused by: java.net.SocketException: Socket is not connected
	at java.net.Socket.getOutputStream(Socket.java:963)
	at redis.clients.jedis.Connection.connect(Connection.java:200)
	... 9 more
```

**Comment by osheroff** [comment]
so https://github.com/xetorthio/jedis/issues/882 seems to indicate that it's actually really no good to try to re-use the same Jedis instance.  Before we go down a deep rabbit hole swapping redis clients, what if we try a smaller patch that simply re-creates the `Jedis` object in the reconnection code?

**Comment by lucastex** [comment]
Great! I'll work on this this weekend and let you know.

**Comment by lucastex** [comment]
Hi @osheroff 

I've started a discussion on this pull request: https://github.com/zendesk/maxwell/pull/1480/files

Thanks,

**Comment by lucastex** [comment]
Hi @osheroff  

We've updated the maxwell docker image in production, and found the same problem. 
But one thing that is intriguing me, is that I've changed the error message in the commit (line 132) here: https://github.com/zendesk/maxwell/pull/1480/files#diff-871ac0c5c38c6144f7761ffa7cb0b17bR132

And after running in production today the error message is the same from the last version:

`lost connection to server, trying to reconnect...` instead of `lost connection to server, recreating redis connection...` present in the commit

```20:43:24,432 WARN  MaxwellRedisProducer - lost connection to server, trying to reconnect...
redis.clients.jedis.exceptions.JedisConnectionException: Attempting to read from a broken connection
	at redis.clients.jedis.Connection.readProtocolWithCheckingBroken(Connection.java:314) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.Connection.getIntegerReply(Connection.java:260) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.Jedis.lpush(Jedis.java:977) ~[jedis-3.1.0.jar:?]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.sendToRedis(MaxwellRedisProducer.java:61) ~[maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.push(MaxwellRedisProducer.java:124) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.processRow(BinlogConnectorReplicator.java:279) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.work(BinlogConnectorReplicator.java:193) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:34) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:224) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.Maxwell.start(Maxwell.java:157) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.Maxwell.main(Maxwell.java:245) [maxwell-1.26.0.jar:1.26.0]
redis.clients.jedis.exceptions.JedisConnectionException: Failed connecting to host xxxxx:6379
	at redis.clients.jedis.Connection.connect(Connection.java:204)
	at redis.clients.jedis.BinaryClient.connect(BinaryClient.java:100)
	at redis.clients.jedis.BinaryJedis.connect(BinaryJedis.java:1866)
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.push(MaxwellRedisProducer.java:130)
	at com.zendesk.maxwell.bootstrap.SynchronousBootstrapper.performBootstrap(SynchronousBootstrapper.java:110)
	at com.zendesk.maxwell.bootstrap.SynchronousBootstrapper.startBootstrap(SynchronousBootstrapper.java:47)
	at com.zendesk.maxwell.bootstrap.BootstrapController.doWork(BootstrapController.java:74)
	at com.zendesk.maxwell.bootstrap.BootstrapController.work(BootstrapController.java:58)
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:34)
	at com.zendesk.maxwell.MaxwellContext.lambda$getBootstrapController$0(MaxwellContext.java:400)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.net.SocketException: Socket is not connected
	at java.base/java.net.Socket.getOutputStream(Socket.java:964)
	at redis.clients.jedis.Connection.connect(Connection.java:200)
	... 10 more```

I'm sure we're running the lastest image, as in our k8s deployment shows:

<img width="515" alt="Screen Shot 2020-05-07 at 18 16 53" src="https://user-images.githubusercontent.com/52998/81345590-1c6f4a00-908f-11ea-8e2c-4318a957e724.png">

Any chances the commit was not included in the version?

**Comment by lucastex** [comment]
Hi @osheroff  

We've updated the maxwell docker image in production, and found the same problem. 
But one thing that is intriguing me, is that I've changed the error message in the commit (line 132) here: https://github.com/zendesk/maxwell/pull/1480/files#diff-871ac0c5c38c6144f7761ffa7cb0b17bR132

And after running in production today the error message is the same from the last version:

`lost connection to server, trying to reconnect...` instead of `lost connection to server, recreating redis connection...` present in the commit

```20:43:24,432 WARN  MaxwellRedisProducer - lost connection to server, trying to reconnect...
redis.clients.jedis.exceptions.JedisConnectionException: Attempting to read from a broken connection
	at redis.clients.jedis.Connection.readProtocolWithCheckingBroken(Connection.java:314) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.Connection.getIntegerReply(Connection.java:260) ~[jedis-3.1.0.jar:?]
	at redis.clients.jedis.Jedis.lpush(Jedis.java:977) ~[jedis-3.1.0.jar:?]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.sendToRedis(MaxwellRedisProducer.java:61) ~[maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.push(MaxwellRedisProducer.java:124) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.processRow(BinlogConnectorReplicator.java:279) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.replication.BinlogConnectorReplicator.work(BinlogConnectorReplicator.java:193) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:34) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.Maxwell.startInner(Maxwell.java:224) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.Maxwell.start(Maxwell.java:157) [maxwell-1.26.0.jar:1.26.0]
	at com.zendesk.maxwell.Maxwell.main(Maxwell.java:245) [maxwell-1.26.0.jar:1.26.0]
redis.clients.jedis.exceptions.JedisConnectionException: Failed connecting to host xxxxx:6379
	at redis.clients.jedis.Connection.connect(Connection.java:204)
	at redis.clients.jedis.BinaryClient.connect(BinaryClient.java:100)
	at redis.clients.jedis.BinaryJedis.connect(BinaryJedis.java:1866)
	at com.zendesk.maxwell.producer.MaxwellRedisProducer.push(MaxwellRedisProducer.java:130)
	at com.zendesk.maxwell.bootstrap.SynchronousBootstrapper.performBootstrap(SynchronousBootstrapper.java:110)
	at com.zendesk.maxwell.bootstrap.SynchronousBootstrapper.startBootstrap(SynchronousBootstrapper.java:47)
	at com.zendesk.maxwell.bootstrap.BootstrapController.doWork(BootstrapController.java:74)
	at com.zendesk.maxwell.bootstrap.BootstrapController.work(BootstrapController.java:58)
	at com.zendesk.maxwell.util.RunLoopProcess.runLoop(RunLoopProcess.java:34)
	at com.zendesk.maxwell.MaxwellContext.lambda$getBootstrapController$0(MaxwellContext.java:400)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.net.SocketException: Socket is not connected
	at java.base/java.net.Socket.getOutputStream(Socket.java:964)
	at redis.clients.jedis.Connection.connect(Connection.java:200)
	... 10 more```

I'm sure we're running the lastest image, as in our k8s deployment shows:

![image](https://user-images.githubusercontent.com/52998/81345753-6eb06b00-908f-11ea-9a6a-19a425c09166.png)

Any chances the commit was not included in the version?

**Comment by osheroff** [comment]
oops, you jumped the gun a bit.  1.26.0 didn't have the fix.  1.26.1 does.

**Comment by lucastex** [comment]
Oh, I'm sorry. Will try and get back with more info later

**Comment by osheroff** [comment]
fixed in 1.26.1.  thanks for sticking with it @lucastex

**Comment by lucastex** [comment]
Everything working great here :) Thanks @osheroff
